state/subnets.go

```diff
@@ -230,6 +232,88 @@ func (s *Subnet) Refresh() error {
 	return nil
 }
 
+// AdditiveUpdate adds new info to the subnet based on provided info.  There are
+// restrictions:
+//   - no change to CIDR, more work to determine how to handle needs to
+// be done.
+//   - no change to ProviderId nor ProviderNetworkID, these are immutable.
+func (s *Subnet) AdditiveUpdate(args SubnetInfo) error {
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if attempt != 0 {
+			if err := s.Refresh(); err != nil {
+				if errors.IsNotFound(err) {
+					return nil, errors.Errorf("ProviderId %q not unique", args.ProviderId)
+				}
+				return nil, errors.Trace(err)
+			}
+		}
+		var ops []txn.Op
+		makeSpaceNameUpdate, err := s.updateSpaceName(args.SpaceName)
+		if err != nil {
+			return nil, err
+		}
+		if makeSpaceNameUpdate {
+			ops = append(ops, txn.Op{
+				C:      subnetsC,
+				Id:     s.doc.DocID,
+				Assert: bson.D{{"txn-revno", s.doc.TxnRevno}},
+				Update: bson.D{{"$set", bson.D{{"space-name", args.SpaceName}}}},
+			})
+		}
+		// TODO (hml) 2019-07-16
+		// Updated once SubnetInfo.AvailablityZone is a slice
+		if s.doc.AvailabilityZone == "" && args.AvailabilityZone != "" {
+			ops = append(ops, txn.Op{
+				C:      subnetsC,
+				Id:     s.doc.DocID,
+				Assert: bson.D{{"txn-revno", s.doc.TxnRevno}},
+				Update: bson.D{{"$set", bson.D{{"availabilityzone", args.AvailabilityZone}}}},
+			})
+		}
+		if s.doc.VLANTag == 0 && args.VLANTag > 0 {
+			ops = append(ops, txn.Op{
+				C:      subnetsC,
+				Id:     s.doc.DocID,
+				Assert: bson.D{{"txn-revno", s.doc.TxnRevno}},
+				Update: bson.D{{"$set", bson.D{{"vlantag", args.VLANTag}}}},
+			})
+		}
+		if len(ops) == 0 {
+			return nil, jujutxn.ErrNoOperations
+		}
+		return ops, nil
+	}
+	return errors.Trace(s.st.db().Run(buildTxn))
+}
+
+func (s *Subnet) updateSpaceName(spaceName string) (bool, error) {
+	var spaceNameChange bool
+	sp, err := s.st.Space(s.doc.SpaceName)
+	switch {
+	case err != nil && !errors.IsNotFound(err):
+		return false, errors.Trace(err)
+	case errors.IsNotFound(err):
+		spaceNameChange = true
+	case err == nil:
+		// TODO (hml) 2019-07-16
+		// Update once network.DefaultSpaceId code has landed.
+		//
+		// The undefined space from MAAS has a providerId of -1.
+		// The juju default space will be 0.
+		spaceNameChange = sp.doc.ProviderId == "-1"
+	}
+	return spaceNameChange && spaceName != "" && s.doc.FanLocalUnderlay == "", nil
+}
+
+// AdditiveSubnetUpdate adds new info to the subnet based on provided info.
+func (st *State) AdditiveSubnetUpdate(args SubnetInfo) error {
```

Comment:
How about just `state.UpdateSubnet` and `subnet.Update`? I know we gate only certain changes, but I think that's fine as a caveat for methods with such names.

Reply:
done